### PR TITLE
Fix Array.prototype.concat() when 'this' is not an array.

### PR DESCRIPTION
**Pull request by dbatyai**
JerryScript-DCO-1.0-Signed-off-by: Dániel Bátyai dbatyai.u-szeged@partner.samsung.com

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp [hunk 1]
```diff
@@ -113,24 +113,40 @@ ecma_builtin_array_prototype_object_concat (ecma_value_t this_arg, /**< this arg
   ecma_completion_value_t new_array = ecma_op_create_array_object (0, 0, false);
   ecma_object_t *new_array_p = ecma_get_object_from_completion_value (new_array);
 
+  if (ecma_object_get_class_name (obj_p) == LIT_MAGIC_STRING_ARRAY_UL)
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp]: @dbatyai, could you, please, add comments with steps from specification that correspond to the operations?

Comment (dbatyai) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp]: @ruben-ayrapetyan, added.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp [hunk 2]
```diff
@@ -113,24 +113,47 @@ ecma_builtin_array_prototype_object_concat (ecma_value_t this_arg, /**< this arg
...
-  for (uint32_t index = 0;
-       index < len && ecma_is_completion_value_empty (ret_value);
-       index++, new_array_index++)
+      /* Using [[Put]] is equvalent to [[DefineOwnProperty]] in this case, so we use it for simplicity. */
+      ecma_completion_value_t put_comp = ecma_op_object_put (new_array_p,
+                                                             index_string_p,
+                                                             get_value,
+                                                             false);
+      JERRY_ASSERT (ecma_is_completion_value_normal (put_comp));
+      ecma_free_completion_value (put_comp);
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp]: As `is_throw` argument of `ecma_op_object_put` is `false`, exceptions can't be thrown and only possible return values in the case are "simple true" or "simple false".
So, we can replace this with corresponding assertion and remove `ecma_free_completion_value` call.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp [hunk 3]
```diff
@@ -113,24 +113,47 @@ ecma_builtin_array_prototype_object_concat (ecma_value_t this_arg, /**< this arg
   ecma_completion_value_t new_array = ecma_op_create_array_object (0, 0, false);
   ecma_object_t *new_array_p = ecma_get_object_from_completion_value (new_array);
 
+  /* 5.b */
+  if (ecma_object_get_class_name (obj_p) == LIT_MAGIC_STRING_ARRAY_UL)
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp]: Seems that the block is very similar to code in the loop below. Could we could move this to a helper and call from both places?

**Comment by dbatyai** [comment]
@ruben-ayrapetyan, I've updated the patch.

File: jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp [hunk 4]
```diff
@@ -326,6 +326,132 @@ ecma_builtin_helper_array_index_normalize (ecma_number_t index, /**< index */
...
+ *
+ * Used by:
+ *         - The Array.prototype.concat routine.
+ *
+ * @return completion value
+ *         Returned value must be freed with ecma_free_completion_value.
+ */
+ecma_completion_value_t
+ecma_builtin_helper_array_concat_value (ecma_object_t *obj_p, /**< array */
+                                        uint32_t *length, /**< in-out: array's length */
```
Comment (ruben-ayrapetyan) [inline jerry-core/ecma/builtin-objects/ecma-builtin-helpers.cpp]: Please, add `_p` suffix.

**Comment by ruben-ayrapetyan** [comment]
@dbatyai, thank you for update.

**Comment by ruben-ayrapetyan** [comment]
After fixing https://github.com/Samsung/jerryscript/pull/415#discussion_r34997412 the changes would look good to me

**Comment by ruben-ayrapetyan** [comment]
Looks good to me

**Comment by galpeter** [comment]
looks good to me

**Comment by egavrin** [comment]
Could you please rebase it to master - I'm getting conflicts at merging.

**Comment by dbatyai** [comment]
@egavrin, rebased to master.

**Pull request merged**